File: spec/lua/test_default_endian_expr_is_be.lua
local luaunit = require("luaunit")

require("default_endian_expr_is_be")

TestDefaultEndianExprIsBe = {}

function TestDefaultEndianExprIsBe:test_default_endian_expr_is_be()
    local r = DefaultEndianExprIsBe:from_file("src/endian_expr.bin")

    -- LE
    luaunit.assertEquals(r.docs[1].indicator, "\x49\x49")
    luaunit.assertEquals(r.docs[1].main.some_int, 0x42)
    luaunit.assertEquals(r.docs[1].main.some_int_be, 0x42)
    luaunit.assertEquals(r.docs[1].main.some_int_le, 0x42)

    luaunit.assertEquals(r.docs[1].main.inst_int, 0x42)
    luaunit.assertEquals(r.docs[1].main.inst_sub.foo, 0x42)

    -- BE
    luaunit.assertEquals(r.docs[2].indicator, "\x4d\x4d")
    luaunit.assertEquals(r.docs[2].main.some_int, 0x42)
    luaunit.assertEquals(r.docs[2].main.some_int_be, 0x42)
    luaunit.assertEquals(r.docs[2].main.some_int_le, 0x42)

    luaunit.assertEquals(r.docs[2].main.inst_int, 0x42000000)
    luaunit.assertEquals(r.docs[2].main.inst_sub.foo, 0x42000000)

    -- Weird => LE
    luaunit.assertEquals(r.docs[3].indicator, "\x58\x58")
    luaunit.assertEquals(r.docs[3].main.some_int, 0x42000000)
    luaunit.assertEquals(r.docs[3].main.some_int_be, 0x42)
    luaunit.assertEquals(r.docs[3].main.some_int_le, 0x42)

    luaunit.assertEquals(r.docs[3].main.inst_int, 0x42)
    luaunit.assertEquals(r.docs[3].main.inst_sub.foo, 0x42)
end

File: spec/java/src/io/kaitai/struct/spec/TestDefaultEndianExprIsBe.java
package io.kaitai.struct.spec;

import io.kaitai.struct.testformats.DefaultEndianExprIsBe;
import org.testng.annotations.Test;

import static org.testng.Assert.assertEquals;

public class TestDefaultEndianExprIsBe extends CommonSpec {
    @Test
    public void testDefaultEndianExprIsBe() throws Exception {
        DefaultEndianExprIsBe r = DefaultEndianExprIsBe.fromFile(SRC_DIR + "endian_expr.bin");

        // LE
        assertEquals(r.docs().get(0).indicator(), new byte[] { 0x49, 0x49 });
        assertEquals(r.docs().get(0).main().someInt(), 0x42);
        assertEquals(r.docs().get(0).main().someIntBe(), 0x42);
        assertEquals(r.docs().get(0).main().someIntLe(), 0x42);

        assertEquals(r.docs().get(0).main().instInt().intValue(), 0x42);
        assertEquals(r.docs().get(0).main().instSub().foo(), 0x42);

        // BE
        assertEquals(r.docs().get(1).indicator(), new byte[] { 0x4d, 0x4d });
        assertEquals(r.docs().get(1).main().someInt(), 0x42);
        assertEquals(r.docs().get(1).main().someIntBe(), 0x42);
        assertEquals(r.docs().get(1).main().someIntLe(), 0x42);

        assertEquals(r.docs().get(1).main().instInt().intValue(), 0x42000000);
        assertEquals(r.docs().get(1).main().instSub().foo(), 0x42000000);

        // Weird => LE
        assertEquals(r.docs().get(2).indicator(), new byte[] { 0x58, 0x58 });
        assertEquals(r.docs().get(2).main().someInt(), 0x42000000);
        assertEquals(r.docs().get(2).main().someIntBe(), 0x42);
        assertEquals(r.docs().get(2).main().someIntLe(), 0x42);

        assertEquals(r.docs().get(2).main().instInt().intValue(), 0x42);
        assertEquals(r.docs().get(2).main().instSub().foo(), 0x42);
    }
}
